### ISSUE TITLE
Error : Value for attr 'data_format' of "", when calling TF_AddGradients with FusedBatchNormV3 operator

### ISSUE THREAD

**Issue by harujoh** [tensorflow/tensorflow#51201]
Hello, I'm very happy to see that FusedBatchNormV3 is now included in backprop.
Unfortunately, when I call TF_AddGradients with a TF_Graph containing a FusedBatchNormV3 operator, I get an error with a blank data_format attribute.

This is probably due to the fact that the default values are not reflected when registering FusedBatchNormGradV3.

**System information**
Windows 10 x64
No custom code
binary From [https://storage.googleapis.com/tensorflow/libtensorflow/libtensorflow-cpu-windows-x86_64-2.6.0.zip](https://storage.googleapis.com/tensorflow/libtensorflow/libtensorflow-cpu-windows-x86_64-2.6.0.zip)
TensorFlow Version 2.6.0
C API 
On CPU

**Comment by harujoh** [comment]
I'm currently waiting for the build, but I think the following fixes will probably fix the problem.

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/cc/gradients/nn_grad.cc#L435-L439

from
  TF_RETURN_IF_ERROR(GetNodeAttr(op.node()->attrs(), "epsilon", &epsilon));  
  TF_RETURN_IF_ERROR(  
      GetNodeAttr(op.node()->attrs(), "data_format", &data_format));  
  TF_RETURN_IF_ERROR(  
      GetNodeAttr(op.node()->attrs(), "is_training", &is_training));  
  
to
  TF_RETURN_IF_ERROR(GetNodeAttr(op.**output(0)**.node()->attrs(), "epsilon", &epsilon));  
  TF_RETURN_IF_ERROR(  
      GetNodeAttr(op.**output(0)**.node()->attrs(), "data_format", &data_format));  
  TF_RETURN_IF_ERROR(  
      GetNodeAttr(op.**output(0)**.node()->attrs(), "is_training", &is_training));

**Comment by harujoh** [comment]
It looks like there are several similar mistakes in this commit.
There are also some suspicious things in the code that existed before this commit.
bf3d89b1e8bc5e00271c2fbd57a8b3ff55d48462

**Comment by sushreebarsa** [comment]
@harujoh In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

**Comment by harujoh** [comment]
Here is a code snippet that reproduces the problem.

It works, but I usually use C#, and I haven't used C++ in 15 years, so please forgive any minor errors.

```
#include <iostream>
#include <tensorflow/c/c_api.h>

int main() {

	auto g = TF_NewGraph();
	auto s = TF_NewStatus();
	int64_t dims[] = { 1, 1, 1, 1 };

	auto ph_desc = TF_NewOperation(g, "Placeholder", "inputVal");
	TF_SetAttrType(ph_desc, "dtype", TF_FLOAT);
	TF_SetAttrShape(ph_desc, "shape", dims, 4);
	auto ph_op = TF_FinishOperation(ph_desc, s);

	auto gamma_desc = TF_NewOperation(g, "Variable", "bn_gamma");
	TF_SetAttrShape(gamma_desc, "shape", nullptr, 0);
	TF_SetAttrType(gamma_desc, "dtype", TF_FLOAT);
	auto gamma_op = TF_FinishOperation(gamma_desc, s);

	auto beta_desc = TF_NewOperation(g, "Variable", "bn_beta");
	TF_SetAttrShape(beta_desc, "shape", nullptr, 0);
	TF_SetAttrType(beta_desc, "dtype", TF_FLOAT);
	auto beta_op = TF_FinishOperation(beta_desc, s);

	auto moving_mean_desc = TF_NewOperation(g, "Variable", "bn_moving_mean");
	TF_SetAttrShape(moving_mean_desc, "shape", dims, 4);
	TF_SetAttrType(moving_mean_desc, "dtype", TF_FLOAT);
	auto moving_mean_op = TF_FinishOperation(moving_mean_desc, s);

	auto moving_var_desc = TF_NewOperation(g, "Variable", "moving_var");
	TF_SetAttrShape(moving_var_desc, "shape", dims, 4);
	TF_SetAttrType(moving_var_desc, "dtype", TF_FLOAT);
	auto moving_var_op = TF_FinishOperation(moving_var_desc, s);

	auto bn_desc = TF_NewOperation(g, "FusedBatchNormV3", "bn");
	TF_Output output;
	output.oper = ph_op;
	output.index = 0;
	TF_AddInput(bn_desc, output);
	output.oper = gamma_op;
	TF_AddInput(bn_desc, output);
	output.oper = beta_op;
	TF_AddInput(bn_desc, output);
	output.oper = moving_mean_op;
	TF_AddInput(bn_desc, output);
	output.oper = moving_var_op;
	TF_AddInput(bn_desc, output);
	TF_SetAttrFloat(bn_desc, "epsilon", 0.0001);
	TF_SetAttrString(bn_desc, "data_format", "NHWC", 4);
	TF_SetAttrBool(bn_desc, "is_training", 1);
	auto bn_op = TF_FinishOperation(bn_desc, s);

	TF_Output bn;
	bn.oper = bn_op;
	bn.index = 0;
	TF_Output outputs[1] = { bn };

	TF_Output gamma;
	gamma.oper = gamma_op;
	gamma.index = 0;
	TF_Output beta;
	beta.oper = beta_op;
	beta.index = 0;
	TF_Output mean;
	mean.oper = moving_mean_op;
	mean.index = 0;
	TF_Output var;
	var.oper = moving_var_op;
	var.index = 0;
	TF_Output variables[4] = { gamma,beta,mean,var };

	TF_Output grad_outputs[1];

	TF_AddGradients(g,
		outputs, 1,
		variables, 4,
		nullptr, s, grad_outputs);

	auto code = TF_GetCode(s);
	if (code != 0) {
		std::cout << TF_Message(s); //Value for attr 'data_format' of "
	}

	return 0;
}
```

**Comment by harujoh** [comment]
I just noticed that you have the wrong version of the tag. 

It is now 2.5, but graphs containing this FusedBatchNormV3 can be used in TF_AddGradients from 2.6rc0.

**Comment by ymodak** [comment]
@harujoh Was this piece of code executing successfully with TF 2.5 for you?

**Comment by harujoh** [comment]
@ymodak I tried with 2.6.0, but I get the same error message.

**Comment by harujoh** [comment]
@ymodak Please change the version tag from 2.6.0-rc0 to 2.6.0 as I have modified the post.

**Comment by zaleslaw** [comment]
The same error on the 2.7.0

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
